apps/script/HoC2023AiGenerateWeights.py

```diff
@@ -107,33 +111,83 @@
+option_lists = [emoji_names_list, palette_model_descriptive_names, background_effect_model_descriptive_names, foreground_effect_model_descriptive_names]
+embeddings = []
+# Final output should be list of lists structured as [[[emoji1], [emoji2], ...], [[palette1]...]...]
+# Where embeddings[0] = all emoji embeddings, [1] = palettes, [2] = background, [3] = foreground
+for i in range (0, 4):
+    embeddings.append([retrieve_embedding(string=item, 
+                                          cache_path=embedding_paths[i],
+                                          embedding_cache=caches[i])
+                                          for item in option_lists[i]])
+
+emoji_embeddings = embeddings[0]
+palette_embeddings = embeddings[1]
+background_embeddings = embeddings[2]
+foreground_embeddings = embeddings[3]
+
+def calculate_similarity_score(input_embeddings, output_embeddings):
+    # Native cosine distance calculation outputs a value between 0 -> 1 where smaller values = greater similarity
+    # We can redefine this into cosine similarity with a simple (x-1)*-1 due to their mathematical relationship
+    # Cosine similarity is preferable as we can easily sum them together to take a max value later
+    similarities = [distances_from_embeddings(input_vector, output_embeddings, distance_metric='cosine')
+                            for input_vector in input_embeddings]
     
-    background_scores = []
-    for background_word in background_effect_model_descriptive_names:
-        id_token = nlp(emoji_name)
-        background_token = nlp(background_word)
-        similarity_score = id_token.similarity(background_token)
-        background_scores.append(round(similarity_score, 2))
-        
-    foreground_scores = []
-    for foreground_word in foreground_effect_model_descriptive_names:
-        id_token = nlp(emoji_name)
-        foreground_token = nlp(foreground_word)
-        similarity_score = id_token.similarity(foreground_token)
-        foreground_scores.append(round(similarity_score, 2))
-
-    emoji_id = emojis_map[emoji_name]
-    palette_dict[emoji_id] = palette_scores
-    background_dict[emoji_id] = background_scores
-    foreground_dict[emoji_id] = foreground_scores
+    # Conversion to pandas DataFrame for ease of manipulation
+    similarities = pd.DataFrame(similarities)
+    similarities.index = emoji_names_list
+    similarities = similarities.apply(lambda x: round((x-1)*-1, 3), axis = 0)
+
+    # Conversion to required JSON lookup format
+    similarities_dict = similarities.transpose().to_dict()
+    for emoji in similarities_dict:
+        similarities_dict[emoji] = list(similarities_dict[emoji].values())
+    return similarities_dict
+
+palette_dict = calculate_similarity_score(emoji_embeddings, palette_embeddings)
+background_dict = calculate_similarity_score(emoji_embeddings, background_embeddings)
+foreground_embeddings = calculate_similarity_score(emoji_embeddings, foreground_embeddings)
```

Comment:
I think maybe you mean to assign `foreground_dict` the result of `calculate_similarity_score` here.